File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/GameServerOnePageResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 请求服务器单组列表(返回) </summary>
public struct GameServerOnePageResponseProto : IProto
{
    public ushort ProtoCode { get { return 50112; } }

    /// <summary> 单组服务器数量 </summary>
    public int ItemCount;
    /// <summary> 单个服务器信息 </summary>
    public List<GameServerOnePageItem> ServerOnePageItemList;

    /// <summary> 单个服务器信息 </summary>
    public struct GameServerOnePageItem
    {
        /// <summary> 服务器Id </summary>
        public int ServerId;
        /// <summary> 服务器状态 </summary>
        public int Status;
        /// <summary> 运行状态 </summary>
        public int RunState;
        /// <summary> 是否推荐 </summary>
        public bool IsCommand;
        /// <summary> 是否是新服 </summary>
        public bool IsNew;
        /// <summary> 服务器名称 </summary>
        public string Name;
        /// <summary> 服务器IP </summary>
        public string Ip;
        /// <summary> 服务器端口号 </summary>
        public int Port;
    }

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(ItemCount);
            for (int i = 0; i < ItemCount; i++)
            {
                ms.WriteInt(ServerOnePageItemList[i].ServerId);
                ms.WriteInt(ServerOnePageItemList[i].Status);
                ms.WriteInt(ServerOnePageItemList[i].RunState);
                ms.WriteBool(ServerOnePageItemList[i].IsCommand);
                ms.WriteBool(ServerOnePageItemList[i].IsNew);
                ms.WriteUTF8String(ServerOnePageItemList[i].Name);
                ms.WriteUTF8String(ServerOnePageItemList[i].Ip);
                ms.WriteInt(ServerOnePageItemList[i].Port);
            }
            return ms.ToArray();
        }
    }

    public static GameServerOnePageResponseProto GetProto(byte[] buffer)
    {
        GameServerOnePageResponseProto proto = new GameServerOnePageResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.ItemCount = ms.ReadInt();
            proto.ServerOnePageItemList = new List<GameServerOnePageItem>();
            for (int i = 0; i < proto.ItemCount; i++)
            {
                GameServerOnePageItem _ServerOnePageItem = new GameServerOnePageItem();
                _ServerOnePageItem.ServerId = ms.ReadInt();
                _ServerOnePageItem.Status = ms.ReadInt();
                _ServerOnePageItem.RunState = ms.ReadInt();
                _ServerOnePageItem.IsCommand = ms.ReadBool();
                _ServerOnePageItem.IsNew = ms.ReadBool();
                _ServerOnePageItem.Name = ms.ReadUTF8String();
                _ServerOnePageItem.Ip = ms.ReadUTF8String();
                _ServerOnePageItem.Port = ms.ReadInt();
                proto.ServerOnePageItemList.Add(_ServerOnePageItem);
            }
        }
        return proto;
    }
}

File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/GameServerOnePageResponseProto.lua
--请求服务器单组列表(返回)
GameServerOnePageResponseProto = { ProtoCode = 50112, ItemCount = 0, ServerOnePageItemTable = { } }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameServerOnePageResponseProto.__index = GameServerOnePageResponseProto;

function GameServerOnePageResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameServerOnePageResponseProto); --将self的元表设定为Class
    return self;
end


--定义单个服务器信息
ServerOnePageItem = { ServerId = 0, Status = 0, RunState = 0, IsCommand = false, IsNew = false, Name = "", Ip = "", Port = 0 }
ServerOnePageItem.__index = ServerOnePageItem;
function ServerOnePageItem.New()
    local self = { };
    setmetatable(self, ServerOnePageItem);
    return self;
end


--发送协议
function GameServerOnePageResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.ItemCount);
    for i = 1, proto.ItemCount, 1 do
        ms:WriteInt(ServerOnePageItemList[i].ServerId);
        ms:WriteInt(ServerOnePageItemList[i].Status);
        ms:WriteInt(ServerOnePageItemList[i].RunState);
        ms:WriteBool(ServerOnePageItemList[i].IsCommand);
        ms:WriteBool(ServerOnePageItemList[i].IsNew);
        ms:WriteUTF8String(ServerOnePageItemList[i].Name);
        ms:WriteUTF8String(ServerOnePageItemList[i].Ip);
        ms:WriteInt(ServerOnePageItemList[i].Port);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameServerOnePageResponseProto.GetProto(buffer)

    local proto = GameServerOnePageResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.ItemCount = ms:ReadInt();
	proto.ServerOnePageItemTable = {};
    for i = 1, proto.ItemCount, 1 do
        local _ServerOnePageItem = ServerOnePageItem.New();
        _ServerOnePageItem.ServerId = ms:ReadInt();
        _ServerOnePageItem.Status = ms:ReadInt();
        _ServerOnePageItem.RunState = ms:ReadInt();
        _ServerOnePageItem.IsCommand = ms:ReadBool();
        _ServerOnePageItem.IsNew = ms:ReadBool();
        _ServerOnePageItem.Name = ms:ReadUTF8String();
        _ServerOnePageItem.Ip = ms:ReadUTF8String();
        _ServerOnePageItem.Port = ms:ReadInt();
        proto.ServerOnePageItemTable[#proto.ServerOnePageItemTable+1] = _ServerOnePageItem;
    end

    ms:Dispose();
    return proto;
end